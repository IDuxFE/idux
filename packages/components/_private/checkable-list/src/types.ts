/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
