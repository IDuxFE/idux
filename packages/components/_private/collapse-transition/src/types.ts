/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
