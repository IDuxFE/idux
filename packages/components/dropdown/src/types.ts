import type { DefineComponent } from 'vue'
import type { OverlayPlacement, OverlayTrigger } from '@idux/cdk/overlay'

import { overlayPlacementDef, overlayTriggerDef } from '@idux/cdk/overlay'
import { PropTypes } from '@idux/cdk/utils'

export interface DropdownProps {
  visible: boolean
  disabled: boolean
  icon?: string
  overlayClass?: string
  placement?: OverlayPlacement
  trigger?: OverlayTrigger
}

export const dropdownPropsDef = {
  visible: PropTypes.bool.def(false),
  disabled: PropTypes.bool.def(false),
  icon: PropTypes.string,
  overlayClass: PropTypes.string,
  placement: overlayPlacementDef,
  trigger: overlayTriggerDef,
}

export type DropdownInstance = InstanceType<DefineComponent<DropdownProps>>

export interface DropdownButtonProps extends DropdownProps {
  icon: string
}

export const dropdownButtonPropsDef = {
  ...dropdownPropsDef,
  icon: PropTypes.string.def('ellipsis'),
}

export type DropdownButtonInstance = InstanceType<DefineComponent<DropdownButtonProps>>
