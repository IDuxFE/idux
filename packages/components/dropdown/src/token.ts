/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { InjectionKey } from 'vue'

export interface DropdownContext {
  setVisibility: (visible: boolean) => void
}

export const dropdownToken: InjectionKey<DropdownContext> = Symbol('dropdownToken')
