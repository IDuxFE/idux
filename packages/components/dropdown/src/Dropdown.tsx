/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, provide, toRef } from 'vue'

import { useControlledProp } from '@idux/cdk/utils'
import { ɵOverlay } from '@idux/components/_private/overlay'
import { useGlobalConfig } from '@idux/components/config'
import { useThemeToken } from '@idux/components/theme'

import { dropdownToken } from './token'
import { dropdownProps } from './types'
import { getThemeTokens } from '../theme'

const defaultDelay: [number, number] = [0, 100]

export default defineComponent({
  name: 'IxDropdown',
  props: dropdownProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const { globalHashId, hashId, registerToken } = useThemeToken('dropdown')
    registerToken(getThemeTokens)

    const config = useGlobalConfig('dropdown')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-dropdown`)

    const [visibility, setVisibility] = useControlledProp(props, 'visible', false)

    provide(dropdownToken, { hideOnClick: toRef(props, 'hideOnClick'), setVisibility })

    return () => {
      const { trigger = config.trigger, placement = config.placement } = props

      const overlayProps = {
        class: [mergedPrefixCls.value, globalHashId.value, hashId.value],
        autoAdjust: props.autoAdjust ?? config.autoAdjust,
        clickOutside: trigger === 'click' || trigger === 'contextmenu',
        container: props.overlayContainer ?? config.overlayContainer,
        containerFallback: `.${mergedPrefixCls.value}-overlay-container`,
        delay: defaultDelay,
        destroyOnHide: props.destroyOnHide ?? config.destroyOnHide,
        disabled: props.disabled,
        offset: props.offset ?? config.offset,
        placement,
        showArrow: props.showArrow ?? config.showArrow,
        transitionName: `${common.prefixCls}-slide-auto`,
        trigger,
        visible: visibility.value,
        ['onUpdate:visible']: setVisibility,
      } as const

      return <ɵOverlay v-slots={{ default: slots.default, content: slots.overlay }} {...overlayProps} />
    }
  },
})
