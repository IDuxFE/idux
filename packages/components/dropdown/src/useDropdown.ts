import type { ComputedRef, Ref } from 'vue'
import type { OverlayElement, OverlayPlacement, OverlayTrigger } from '@idux/cdk/overlay'
import type { DropdownProps } from './types'

import { computed, getCurrentInstance, onMounted, provide, ref, watch } from 'vue'
import { useOverlay } from '@idux/cdk/overlay'
import { useGlobalConfig } from '@idux/components/config'
import { dropdownToken } from './token'

export interface UseDropdownConfig {
  placement: ComputedRef<OverlayPlacement>
  trigger: ComputedRef<OverlayTrigger>
}

export const useDropdownConfig = (props: DropdownProps): UseDropdownConfig => {
  const config = useGlobalConfig('dropdown')
  const placement = computed(() => props.placement ?? config.placement)
  const trigger = computed(() => props.trigger ?? config.trigger)
  return { placement, trigger }
}

export interface UseDropdownOpenState {
  openState: Ref<boolean>
  onMouseOverlayChang: (open: boolean) => void
}

export const useDropdownOpenState = (trigger: ComputedRef<OverlayTrigger>): UseDropdownOpenState => {
  const openState = ref(false)
  const onMouseOverlayChang = (open: boolean) => {
    if (trigger.value === 'hover') {
      openState.value = open
    }
  }

  const setMenuOpenState = (open: boolean) => {
    openState.value = open
  }

  provide(dropdownToken, { setMenuOpenState, onMouseOverlayChang })

  return { openState, onMouseOverlayChang }
}

export interface UseDropdownOverlay {
  triggerRef: Ref<OverlayElement | null>
  overlayRef: Ref<OverlayElement | null>
  visibility: ComputedRef<boolean>
  onClickOutside: () => void
}

export const useDropdownOverlay = (
  props: DropdownProps,
  placement: ComputedRef<OverlayPlacement>,
  trigger: ComputedRef<OverlayTrigger>,
  openState: Ref<boolean>,
): UseDropdownOverlay => {
  const { triggerRef, overlayRef, visibility, show, hide, update, initialize } = useOverlay({
    scrollStrategy: 'reposition',
    placement: placement.value,
    trigger: trigger.value,
    visible: props.visible,
    offset: [0, 8],
    showDelay: 0,
    hideDelay: 300,
  })

  onMounted(() => initialize())

  watch([placement, trigger], ([currPlacement, currTrigger]) => {
    update({ placement: currPlacement, trigger: currTrigger })
  })

  watch(
    () => props.visible,
    visible => {
      if (visible !== visibility.value) {
        visible ? show() : hide()
      }
    },
  )

  watch(openState, currOpenState => {
    currOpenState ? show() : hide()
  })

  const { emit } = getCurrentInstance()!
  watch(visibility, visible => {
    emit('update:visible', visible)
  })

  const onClickOutside = () => {
    if (trigger.value === 'click' || trigger.value === 'contextmenu') {
      hide()
    }
  }

  return { triggerRef, overlayRef, visibility, onClickOutside }
}
