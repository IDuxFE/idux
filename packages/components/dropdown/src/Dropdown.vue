<template>
  <a ref="triggerRef" v-click-outside="onClickOutside" class="ix-dropdown">
    <slot />
    <slot name="icon"><ix-icon v-if="icon" :name="icon" /></slot>
    <ix-portal target="ix-dropdown-container">
      <transition>
        <div
          v-show="visibility"
          ref="overlayRef"
          :class="overlayClass"
          @mouseenter="onMouseOverlayChang(true)"
          @mouseleave="onMouseOverlayChang(false)"
        >
          <slot name="overlay" />
        </div>
      </transition>
    </ix-portal>
  </a>
</template>

<script lang="ts">
import type { DropdownProps } from './types'

import { defineComponent } from 'vue'
import { clickOutside } from '@idux/cdk/click-outside'
import { IxPortal } from '@idux/cdk/portal'
import { IxIcon } from '@idux/components/icon'
import { dropdownPropsDef } from './types'
import { useDropdownConfig, useDropdownOpenState, useDropdownOverlay } from './useDropdown'

export default defineComponent({
  name: 'IxDropdown',
  components: { IxPortal, IxIcon },
  directives: { clickOutside },
  props: dropdownPropsDef,
  emits: ['update:visible'],
  setup(props: DropdownProps) {
    const { placement, trigger } = useDropdownConfig(props)
    const { openState, onMouseOverlayChang } = useDropdownOpenState(trigger)
    const { triggerRef, overlayRef, visibility, onClickOutside } = useDropdownOverlay(
      props,
      placement,
      trigger,
      openState,
    )

    return { triggerRef, overlayRef, visibility, onClickOutside, onMouseOverlayChang }
  },
})
</script>
