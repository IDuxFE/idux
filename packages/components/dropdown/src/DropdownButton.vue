<template>
  <ix-button-group v-click-outside="onClickOutside" class="ix-dropdown">
    <ix-button><slot /></ix-button>
    <ix-button ref="triggerRef">
      <slot name="icon"><ix-icon :name="icon" /></slot>
    </ix-button>
    <ix-portal target="ix-dropdown-container">
      <transition>
        <div
          v-show="visibility"
          ref="overlayRef"
          :class="overlayClass"
          @mouseenter="onMouseOverlayChang(true)"
          @mouseleave="onMouseOverlayChang(false)"
        >
          <slot name="overlay" />
        </div>
      </transition>
    </ix-portal>
  </ix-button-group>
</template>

<script lang="ts">
import type { DropdownButtonProps } from './types'

import { defineComponent } from 'vue'
import { clickOutside } from '@idux/cdk/click-outside'
import { IxPortal } from '@idux/cdk/portal'
import { IxButton, IxButtonGroup } from '@idux/components/button'
import { IxIcon } from '@idux/components/icon'
import { dropdownButtonPropsDef } from './types'
import { useDropdownConfig, useDropdownOpenState, useDropdownOverlay } from './useDropdown'

export default defineComponent({
  name: 'IxDropdownButton',
  components: { IxPortal, IxIcon, IxButton, IxButtonGroup },
  directives: { clickOutside },
  props: dropdownButtonPropsDef,
  emits: ['update:visible'],
  setup(props: DropdownButtonProps) {
    const { placement, trigger } = useDropdownConfig(props)
    const { openState, onMouseOverlayChang } = useDropdownOpenState(trigger)
    const { triggerRef, overlayRef, visibility, onClickOutside } = useDropdownOverlay(
      props,
      placement,
      trigger,
      openState,
    )

    return { triggerRef, overlayRef, visibility, onClickOutside, onMouseOverlayChang }
  },
})
</script>
