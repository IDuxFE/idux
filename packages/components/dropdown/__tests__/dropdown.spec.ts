import { mount, MountingOptions, VueWrapper } from '@vue/test-utils'
import IxDropdown from '../src/Dropdown.vue'
import { DropdownProps } from '../src/types'

describe('Dropdown.vue', () => {
  let DropdownMount: (options?: MountingOptions<Partial<DropdownProps>>) => VueWrapper<InstanceType<typeof IxDropdown>>

  beforeEach(() => {
    DropdownMount = options => mount(IxDropdown, { ...options })
  })

  test('render work', () => {
    const wrapper = DropdownMount({})

    expect(wrapper.html()).toMatchSnapshot()

    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.unmount()
    }).not.toThrow()
  })
})
