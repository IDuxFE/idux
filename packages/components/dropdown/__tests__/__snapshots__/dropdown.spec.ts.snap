// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown.vue render work 1`] = `
"<a class=\\"ix-dropdown\\">
  <!--v-if-->
  <ix-portal target=\\"ix-dropdown-container\\">
    <transition-stub>
      <div class=\\"\\" style=\\"display: none; position: absolute; left: 0px; top: 0px; margin: 0px;\\"></div>
    </transition-stub>
  </ix-portal>
</a>"
`;
